### PR TITLE
read: fix panic on witness check

### LINKED ISSUE #13764: tikv panic when inject network fault
### What version of TiKV are you using?
./tikv-server -V
 TiKV 
Release Version:   6.5.0-alpha
Edition:           Community
Git Commit Hash:   f2e89a4e80e2b99d7ac2abe74d38d4f6eac9ceb6
Git Commit Branch: heads/refs/tags/v6.5.0-alpha
UTC Build Time:    2022-11-09 11:07:18
Rust Version:      rustc 1.64.0-nightly (0f4bcadb4 2022-07-30)
Enable Features:   pprof-fp jemalloc mem-profiling portable sse test-engine-kv-rocksdb test-engine-raft-raft-engine cloud-aws cloud-gcp cloud-azure
Profile:           dist_release
2022-11-10T09:52:49.633+0800	INFO	k8s/client.go:131	it should be noted that a long-running command will not be interrupted even the use case has ended. For more information, please refer to https://github.com/pingcap/test-infra/discussions/129

### What operating system and CPU are you using?
8c、32g

### Steps to reproduce
https://tcms.pingcap.net/dashboard/executions/plan/1269113

### What did you expect?
no panic

### What did happened?
tikv panic
[2022/11/10 02:22:22.382 +00:00] [FATAL] [lib.rs:495] ["called `Option::unwrap()` on a `None` value"] [backtrace="   0: tikv_util::set_panic_hook::{{closure}}\n             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/tikv_util/src/lib.rs:494:18\n   1: std::panicking::rust_panic_with_hook\n             at rust/toolchains/nightly-2022-07-31-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:702:17\n   2: std::panicking::begin_panic_handler::{{closure}}\n             at rust/toolchains/nightly-2022-07-31-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:586:13\n   3: std::sys_common::backtrace::__rust_end_short_backtrace\n             at rust/toolchains/nightly-2022-07-31-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:138:18\n   4: rust_begin_unwind\n             at rust/toolchains/nightly-2022-07-31-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:584:5\n   5: core::panicking::panic_fmt\n             at rust/toolchains/nightly-2022-07-31-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panicking.rs:142:14\n   6: core::panicking::panic\n             at rust/toolchains/nightly-2022-07-31-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panicking.rs:48:5\n   7: core::option::Option<T>::unwrap\n             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/raftstore/src/store/worker/read.rs:789:12\n      raftstore::store::worker::read::LocalReaderCore<D,S>::validate_request\n             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/raftstore/src/store/worker/read.rs:789:12\n   8: raftstore::store::worker::read::LocalReader<E,C>::pre_propose_raft_command\n             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/raftstore/src/store/worker/read.rs:864:33\n      raftstore::store::worker::read::LocalReader<E,C>::propose_raft_command\n             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/raftstore/src/store/worker/read.rs:918:15\n      raftstore::store::worker::read::LocalReader<E,C>::read\n             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/raftstore/src/router.rs:251:9\n      <raftstore::router::ServerRaftStoreRouter<EK,ER> as raftstore::router::LocalReadRouter<EK>>::read\n             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/raftstore/src/router.rs:251:9\n      tikv::server::raftkv::RaftKv<E,S>::exec_snapshot\n             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/src/server/raftkv.rs:221:9\n      <tikv::server::raftkv::RaftKv<E,S> as tikv_kv::Engine>::async_snapshot\n             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/src/server/raftkv.rs:450:9\n   9: tikv_kv::snapshot\n             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/tikv_kv/src/lib.rs:592:15\n  10: tikv::storage::Storage<E,L,F>::snapshot\n             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/src/storage/mod.rs:325:9\n      tikv::storage::Storage<E,L,F>::get::{{closure}}::{{closure}}\n             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/src/storage/mod.rs:639:52\n      tikv_kv::with_tls_engine::{{closure}}\n             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/tikv_kv/src/lib.rs:545:9\n      std::thread::local::LocalKey<T>::try_with\n             at rust/toolchains/nightly-2022-07-31-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/local.rs:445:16\n      std::thread::local::LocalKey<T>::with\n             at rust/toolchains/nightly-2022-07-31-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/local.rs:421:9\n      tikv_kv::with_tls_engine\n             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/tikv_kv/src/lib.rs:543:5\n      tikv::storage::Storage<E,L,F>::with_tls_engine\n             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/src/storage/mod.rs:358:18\n      tikv::storage::Storage<E,L,F>::get::{{closure}}\n             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/src/storage/mod.rs:639:21\n      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll\n             at rust/toolchains/nightly-2022-07-31-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/mod.rs:91:19\n      <resource_metering::InTags<T> as core::future::future::Future>::poll\n             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/resource_metering/src/lib.rs:266:9\n      tikv::read_pool::ReadPoolHandle::spawn_handle::{{closure}}\n             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/src/read_pool.rs:170:28\n      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll\n             at rust/toolchains/nightly-2022-07-31-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/mod.rs:91:19\n  11: tikv::read_pool::ReadPoolHandle::spawn::{{closure}}\n             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/src/read_pool.rs:146:26\n      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll\n             at rust/toolchains/nightly-2022-07-31-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/mod.rs:91:19\n      <tracker::tls::TrackedFuture<F> as core::future::future::Future>::poll::{{closure}}\n             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/tracker/src/tls.rs:64:23\n      std::thread::local::LocalKey<T>::try_with\n             at rust/toolchains/nightly-2022-07-31-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/local.rs:445:16\n      std::thread::local::LocalKey<T>::with\n             at rust/toolchains/nightly-2022-07-31-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/local.rs:421:9\n      <tracker::tls::TrackedFuture<F> as core::future::future::Future>::poll\n             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/tracker/src/tls.rs:62:9\n      yatp::task::future::RawTask<F>::poll\n             at rust/git/checkouts/yatp-e704b73c3ee279b6/39cb495/src/task/future.rs:59:9\n  12: yatp::task::future::TaskCell::poll\n             at rust/git/checkouts/yatp-e704b73c3ee279b6/39cb495/src/task/future.rs:103:9\n      <yatp::task::future::Runner as yatp::pool::runner::Runner>::handle\n             at rust/git/checkouts/yatp-e704b73c3ee279b6/39cb495/src/task/future.rs:387:20\n  13: <tikv_util::yatp_pool::YatpPoolRunner<T> as yatp::pool::runner::Runner>::handle\n             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/tikv_util/src/yatp_pool/mod.rs:122:24\n      <yatp::queue::multilevel::MultilevelRunner<R> as yatp::pool::runner::Runner>::handle\n             at rust/git/checkouts/yatp-e704b73c3ee279b6/39cb495/src/queue/multilevel.rs:273:19\n      yatp::pool::worker::WorkerThread<T,R>::run\n             at rust/git/checkouts/yatp-e704b73c3ee279b6/39cb495/src/pool/worker.rs:48:13\n      yatp::pool::builder::LazyBuilder<T>::build::{{closure}}\n             at rust/git/checkouts/yatp-e704b73c3ee279b6/39cb495/src/pool/builder.rs:114:25\n      std::sys_common::backtrace::__rust_begin_short_backtrace\n             at rust/toolchains/nightly-2022-07-31-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:122:18\n  14: std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}\n             at rust/toolchains/nightly-2022-07-31-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/mod.rs:505:17\n      <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once\n             at rust/toolchains/nightly-2022-07-31-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panic/unwind_safe.rs:271:9\n      std::panicking::try::do_call\n             at rust/toolchains/nightly-2022-07-31-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:492:40\n      std::panicking::try\n             at rust/toolchains/nightly-2022-07-31-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:456:19\n      std::panic::catch_unwind\n             at rust/toolchains/nightly-2022-07-31-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:137:14\n      std::thread::Builder::spawn_unchecked_::{{closure}}\n             at rust/toolchains/nightly-2022-07-31-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/mod.rs:504:30\n      core::ops::function::FnOnce::call_once{{vtable.shim}}\n             at rust/toolchains/nightly-2022-07-31-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:248:5\n  15: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once\n             at rust/toolchains/nightly-2022-07-31-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:1935:9\n      <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once\n             at rust/toolchains/nightly-2022-07-31-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:1935:9\n      std::sys::unix::thread::Thread::new::thread_start\n             at rust/toolchains/nightly-2022-07-31-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys/unix/thread.rs:108:17\n  16: <unknown>\n  17: clone\n"] [location=/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/raftstore/src/store/worker/read.rs:789] [thread_name=unified-read-pool-3]
[2022/11/10 02:22:25.327 +00:00] [INFO] [lib.rs:85] ["Welcome to TiKV"]

### PR DESCRIPTION
**Pull request by Connor1996**
Signed-off-by: Connor1996 <zbk602423539@gmail.com>

### What is changed and how it works?

Issue Number: Close #13764 

What's Changed:

```commit-message
After the remove peer conf-change is applied and before the peer is destroyed. 
There is a chance that local reader may not find the peer from region info. 
So this PR considers this case and fixes panic on witness check.
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- No code

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/raftstore/src/store/worker/read.rs [hunk 1]
```diff
@@ -813,6 +804,12 @@ where
             return Ok(None);
         }
 
+        // Check witness
+        if find_peer_by_id(&delegate.region, delegate.peer_id).map_or(true, |p| p.is_witness) {
```
Comment (BusyJay) [inline components/raftstore/src/store/worker/read.rs:808]: Any test case?

Comment (Connor1996) [reply components/raftstore/src/store/worker/read.rs:808]: It's very straightforward to avoid the unwrap panic, don't think it's necessary

Comment (BusyJay) [reply components/raftstore/src/store/worker/read.rs:808]: If it's that straightforward, why it's not avoided in the last PR?

Comment (BusyJay) [reply components/raftstore/src/store/worker/read.rs:808]: Why is the default true?

Comment (Connor1996) [reply components/raftstore/src/store/worker/read.rs:808]: Not sure whether it is a witness, better reject it

Comment (BusyJay) [reply components/raftstore/src/store/worker/read.rs:808]: Technically if epoch matches then it must be searchable. So `unwrap` is actually OK here.

**Review by BusyJay — APPROVED** [review]

**Pull request merged**